### PR TITLE
[SignalR] Pass new App to Transport Pipe to App layer

### PR DESCRIPTION
**Pull request by BrennanConroy**
**Pass new 'App to Transport Pipe' to 'App layer'**

Removes the use of `ConnectionResetException` to let the app layer know the connection is being replaced. Instead we pass the new `Pipe` to the App layer so it knows when a new underlying connection is present.

Also, adds `DisableReconnect` so the app layer can let the connection layer know that new connections shouldn't be allowed anymore.

### CODE REVIEW

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs
Comment (halter73) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:583]: Does this mean that if we do make API changes to preview features in 9.0, it could make things break harder when mixing 8.0 and 9.0 packages?

File: src/SignalR/common/Shared/MessageBuffer.cs
Comment (halter73) [inline src/SignalR/common/Shared/MessageBuffer.cs:229]: Nit:

```suggestion
    internal async Task ResendAsync(PipeWriter writer)
```

And then delete `DoResendAsync` and just have this method body continue. That part is almost simple enough to suggest in a diff, but it doesn't work since it spans unchanged lines.

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs
Comment (halter73) [inline src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs:682]: Doesn't it feel like this should be synchronized? Same with the variant in `HttpConnectionContext`. What if we fire the `notifyOnReconnect` event way after `DisableReconnect()` returns? Potentially after the transport and/or connection have fully closed?

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs
Comment (halter73) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:222]: Do we really want to await resending everything before disabling the request body timeout? It seems like that could cause us to close healthy connections. Is there any possible benefit to awaiting this task immediately rather than putting it in the WhenAny?

The same goes on the client side even though there's no IHttpRequestTimeoutFeature. Why would the implementor of NotifyOnReconnect ever want to delay execution of anything at the Http.Connections layer? What's the `reconnectTask` for? @davidfowl

**Review by halter73 — APPROVED** [review]

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs
Comment (BrennanConroy) [reply src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:583]: So like the following examples:
* If you used 8.0 client on the server that targets 9.0
* If you use Connection.Abstractions 9.0 with 8.0 client

Yeah that does seem bad :/
Generally we say "use the same version packages" since we ship everything at the same time. But it's sort of a cop-out.

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs
Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:222]: > Is there any possible benefit to awaiting this task immediately rather than putting it in the WhenAny?

Well the `WhenAny` is for when the connection is closing, either due to the transport detecting a disconnect or the application completing. So we'd still want to wait for the reconnect task separately. But we should put that after the IHttpRequestTimeoutFeature.

File: src/Servers/Connections.Abstractions/src/Features/IStatefulReconnectFeature.cs
Comment (davidfowl) [inline src/Servers/Connections.Abstractions/src/Features/IStatefulReconnectFeature.cs:28]:
```suggestion
    public void OnReconnected(Func<PipeWriter, Task> callback);
```

File: src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs
Comment (davidfowl) [inline src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs:72]: dead code?

**Comment by adityamandaleeka** [comment]
This came in before 5 and will get merged, but I'm waiting for a build to succeed after the latest merge conflict resolution.

**Pull request merged**
